src/typing_extensions.py

```diff
@@ -2337,21 +2337,30 @@ def decorator(arg: _T, /) -> _T:
                 return arg
             elif isinstance(arg, type):
                 original_new = arg.__new__
-                has_init = arg.__init__ is not object.__init__
+                original_init_subclass = arg.__init_subclass__
 
                 @functools.wraps(original_new)
                 def __new__(cls, *args, **kwargs):
-                    warnings.warn(msg, category=category, stacklevel=stacklevel + 1)
+                    if cls is arg:
+                        warnings.warn(msg, category=category, stacklevel=stacklevel + 1)
                     if original_new is not object.__new__:
                         return original_new(cls, *args, **kwargs)
                     # Mirrors a similar check in object.__new__.
-                    elif not has_init and (args or kwargs):
+                    elif cls.__init__ is object.__init__ and (args or kwargs):
                         raise TypeError(f"{cls.__name__}() takes no arguments")
                     else:
                         return original_new(cls)
 
                 arg.__new__ = staticmethod(__new__)
+
+                @functools.wraps(original_init_subclass)
+                def __init_subclass__(*args, **kwargs):
+                    warnings.warn(msg, category=category, stacklevel=stacklevel + 1)
+                    return original_init_subclass(*args, **kwargs)
+
+                arg.__init_subclass__ = __init_subclass__
                 arg.__deprecated__ = __new__.__deprecated__ = msg
+                __init_subclass__.__deprecated__ = msg
```

Comment:
This doesn't always do the right thing, because of the fact that `__init_subclass__` is an implicit classmethod. For example:

```pycon
Python 3.11.5 (tags/v3.11.5:cce6ba9, Aug 24 2023, 14:38:34) [MSC v.1936 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing_extensions as t
>>> @t.deprecated('DEPRECATION')
... class Foo:
...     ATTR = 1
...     def __init_subclass__(cls, *args, **kwargs):
...         print(cls.ATTR)
...
>>> class Bar:
...     ATTR = 2
...
>>> class Baz(Bar, Foo): pass
...
<stdin>:1: DeprecationWarning: DEPRECATION
1
>>> Baz.ATTR
2
```

You can fix it by doing this (diff is relative to your PR branch):

```diff
diff --git a/src/typing_extensions.py b/src/typing_extensions.py
index 6a0015a..37ef575 100644
--- a/src/typing_extensions.py
+++ b/src/typing_extensions.py
@@ -2337,7 +2337,7 @@ else:
                 return arg
             elif isinstance(arg, type):
                 original_new = arg.__new__
-                original_init_subclass = arg.__init_subclass__
+                original_init_subclass = arg.__init_subclass__.__func__

                 @functools.wraps(original_new)
                 def __new__(cls, *args, **kwargs):
@@ -2358,7 +2358,7 @@ else:
                     warnings.warn(msg, category=category, stacklevel=stacklevel + 1)
                     return original_init_subclass(*args, **kwargs)

-                arg.__init_subclass__ = __init_subclass__
+                arg.__init_subclass__ = classmethod(__init_subclass__)
                 arg.__deprecated__ = __new__.__deprecated__ = msg
                 __init_subclass__.__deprecated__ = msg
                 return arg
```

Reply:
Thanks for catching this. Your solution isn't quite right as it doesn't work for classes without a Python `__init_subclass__`, but I got something to work.

Reply:
Ah right, thanks -- I checked my solution worked in the REPL, but didn't actually run the test suite with my changes :) My bad!